### PR TITLE
Add tests for the `browsingContext.fragmentNavigated` event

### PR DESCRIPTION
**Pull request by OrKoN**
- [x] `./wpt lint` passes
- [x] tests pass with chromium-bidi
- [x] history API navigations
- [x] regular navigation
- [x] test_subscribe
- [x] test_unsubscribe
- [x] test_timestamp
- [x] test with iframes
- [x] test with document.write
- [x] test with a base tag

### CODE REVIEW

File: webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py
Comment (OrKoN) [inline webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py:34]: should I perhaps parametrize further? will probably do that tomorrow

Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py:34]: I mean smth like

```
@pytest.mark.parametrize(
    "navigationType",
   ['browsing_context.navigate', 'script.call_function'],
    ids=['BiDi navigation', 'script evaluation'],
)
```

Comment (thiagowfx) [reply webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py:34]: Yes!

**Comment by OrKoN** [comment]
It looks like we also need to check these events for regular navigations.

**Comment by whimboo** [comment]
> It looks like we also need to check these events for regular navigations.

I assume it means that you will add more tests and that a review should happen once the update has been pushed?

**Comment by OrKoN** [comment]
@whimboo if you have some other ideas for which tests we would need, feel free to comment (I will add the regular navigation tests tomorrow)

**Comment by OrKoN** [comment]
Added a few tests (should be ready for review)

**Comment by OrKoN** [comment]
@thiagowfx PTAL

File: webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py
Comment (thiagowfx) [inline webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py:101]: This will never happen, as we're providing `navigation_kind`. I'd just omit the `else`.

**Review by thiagowfx — APPROVED** [review]

File: webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py:10]: Please rebase to latest master given that https://github.com/web-platform-tests/wpt/pull/40134 moved those support files one level up.

File: webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py:14]: Mind following the style of tests that we already have for events? Have a look at the `domContentLoaded` tests as example. Those give a great base for what needs to be tested in general.

**Review by whimboo — CHANGES_REQUESTED** [review]

File: webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py:14]: Could you be more specific about what you mean by this (`./wpt lint` passes so I assume all style considerations are passing)? or do you mean test_unsubscribe/test_subscribe? I don't think it's worthwhile testing this for every event as those should be handled by the session module?

Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py:14]: yeah it's not about linting sorry. But what they test. Also the subscription and unsubscription we need to test that for each event given that each module has to run special steps and implementations might not behave correctly. Within the session module we test all the different argument combinations.

Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py:14]: do we want to test relative order of events? E.g., that fragmentNavigated is emitted before domContentLoaded and after contextCreated? Would it make sense to group event tests by the operation they are triggered by?

Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/navigate/fragment_navigated.py]: added iframe and new context tests. I think that should cover the main cases?

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:157]: Lets keep the name as what we already have for the other test.

```suggestion
async def test_navigate_in_the_same_document(
```

Also we do not specifically test the history API here.

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:140]: Good addition. I think the other navigation related events could also benefit from that. Might be a good follow-up. Maybe you could file at least an issue for wpt?

But I wonder if we should have them as two separate tests without parameterization. Then we could actually parameterize the `test_script` test for `callFunction` and `evaluate`.

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:185]: Shouldn't we just use `document.location = "<new_url>"` to trigger a navigation for the evaluation tests? The History API specific tests I would propose to move to its own file given that it might become larger over time. Right now we only have a basic test.

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:224]: I assume that this event also needs to be fired for `document.write` usage? Given that we had a problem in the past we should potentially add that as well. This would also include the test for a base tag which both other events already have.

**Review by whimboo — CHANGES_REQUESTED** [review]
Thanks Alex for the PR. I had a look at it and have a couple of comments that you can find inline. Not all are blockers but we should have at least the same coverage as the other existing navigation events.

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:127]: previous tests don't verify that the result navigation ID matches the event's navigation ID (and specifically for the navigation that results in a new document being created for the context).

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:224]: I think that there should be no navigation event for document.write usage. I will add a test to check that. As for the base tag, I am not sure how it related to the navigation event? Was it added because of an issue in the Firefox implementation?

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:185]: yeah, we could use document.location I think.

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:223]: As we discussed some months ago and also implemented in Firefox the `document.write` usage should trigger the `DOMContentLoaded` and `load` events. So not sure if I understand your reply. Did you only refer to the `browsingContext.fragmentNavigated` event?

Yes, for URLs with a base tag we incorrectly reported the URL. Don't you think that such a test is helpful for other browsers as well?

Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:223]: Yes, I mean that `document.write` should not trigger `browsingContext.fragmentNavigated` and I have added a test. I have also added the base tag test yesterday (thanks for clarifying that the test exists due to a bug).

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:127]: As such could we rename the test to `test_navigation_id` maybe? I think that's better understandable as `test_cross_document`. Btw. that would be a great test as well for the other navigation events.

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:140]: If you don't want to split this into two distinct tests, can you please at least add id's here so that we have a proper naming of the generated test names? Thanks.

I would still suggest to have two distinct tests given that we do not have a navigation id anyway for the scripting case.

Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:140]: I filed https://github.com/web-platform-tests/wpt/issues/40866 To me it seems like have one test makes sense to avoid duplicating the hashes (originally I had two tests for this but decided to merge it)

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:223]: Ok, so I'm still a bit baffled and sorry for mentioning that so late. In the above tests (like `test_timestamp`) which exercise a normal navigation from eg. domain A to domain B the tests also expect a `fragmentNavigated` event. Why that? As per spec the event should only be fired when a navigation within the same page is happening but those navigations don't qualify for that.

Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:223]: Ah, good catch! Apparently it is a bug in the implementation! I will update the test.

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:97]: Updated

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:288]: Updated the test and added a test that regular navigations should not emit this event.

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:145]: I feel with three levels of parameterization we clearly end-up with a matrix where a lot of tests are not needed and we waste time in running them.

As such I would really propose the following:

* There is no need to test everything for iframes. A single test here is really totally enough and that has to ensure that the contained `context` is the iframe's browsing context.
* Parameterization is fine when different values should be used for a given variable etc, but having huge code blocks behind doesn't really serve the purpose. As such lets really divide it up into three individual tests and each of them parameterize the hashes. Then we may not even need the ids given that a name like `test_document_location[#foo-#bar]` looks fine and is understandable.

Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:208]: alright, I have split the tests. PTAL.

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:131]: nit: other tests navigate from no hash to `#foo` so this test is a bit different, but would be fine.

Thanks for splitting it off! It saves us quite some time in this test.

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (whimboo) [inline webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:167]: Actually I missed that particular test. Given that the fragment is not changing at all, this should trigger the event right?

Same would apply to the other tests as well.

Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:167]: did you mean `this should NOT trigger the event`? my current understanding is that it should be let me consult the spec for this.

Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:167]: I don't see an early return that would prevent event from firing https://html.spec.whatwg.org/multipage/browsing-the-web.html#scroll-to-fragid

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (OrKoN) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:131]: yeah, I am not sure where the balance that we want to strike lies (I have reduced this test under assumption that the navigation to fragment would be similarly handled in iframes vs top-level). We could, of course, change it to include all hashes (and all ways to navigate?) but in practice there could be many more edge cases which are far more likely (like missing events from cross-origin iframes, fenced frames etc). Not sure how far we should go, so defer to your guidance here.

Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:131]: No we most likely wont have to check all the variants here. I only wanted to mention that this test is unique given that is uses a different set of urls. But nothing to worry and we can close this issue if you are fine.

File: webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py
Comment (whimboo) [reply webdriver/tests/bidi/browsing_context/fragment_navigated/fragment_navigated.py:167]: As discussed on Matrix @juliandescottes mentioned that internally we also get a `LOCATION_CHANGE_SAME_DOCUMENT` notification when navigating to the same fragment. Also the spec isn't that much clear, so it looks like that it is fine to keep this test.

**Review by whimboo — APPROVED** [review]
Thanks! This looks good to me now. 

@thiagowfx mind another final review from Google's side?

**Comment by thiagowfx** [comment]
@Lightning00Blade will review from the google side as he's more familiar with this than I am. I'll take another look anyway but please wait for his review before merging.

**Review by thiagowfx — APPROVED** [review]

**Review by Lightning00Blade — APPROVED** [review]

**Comment by thiagowfx** [comment]
@jgraham this can be merged now.

**Comment by whimboo** [comment]
@OrKoN thanks again for writing the tests!

**Pull request merged**
